### ISSUE TITLE
Feature Request: Archiving Emails

### ISSUE THREAD

**Issue by ghost** [foundation/foundation-emails#470]
**Problem**
While working with a team of designers, we are constantly adding to the pages directory for a particular brands emails. We send out twice a week and sometimes three times a week. At the end of the year, this will mean there are lots of emails that get inlined everytime the foundation build command is run. This may cause performance issue after awhile. 

**Proposing:**
A way to archive / ignore files so gulp doesn't inline them every time. @rafibomb had suggested creating a variable in front matter like so: 

```

---
archive: true

---
```

Then the pages function in the gulpfile will not copy those over.

**Comment by rafibomb** [comment]
The reason I thought that might be good, is that the inlining task runs in the dist folder. So if we copy over only the items we want to inline smartly, the overhead will be lower.

I ran it by @kball, he suggested that perhaps the time stamp can be used to process only files that show changes recently. It should also have an option to force all.

This is how Gulp Watch works, sees a change registered in a file or files and runs a task on those. This might be another smarter method.

**Comment by ben-w-smith** [comment]
Needed this functionality aswell @coreyschaaf. The archive = true looked like a good solution so here is my implementation:

You'll need to run two npm installs:

```
$ npm install yaml-front-matter 
// turns yfm into json
$ npm install gulp-ignore
// allows you to ignore a file in the glob based on a boolean
```

Then change the pages function to this:

```
function pages() {
  return gulp.src('src/pages/**/*.html')
    .pipe($.ignore.exclude( function (file) {
        // parse the file and get a json
        var yfmJson = yfm.loadFront(file.contents.toString('ascii'));
        // checks if json has archive key and if its value is true
        var archive = yfmJson.hasOwnProperty('archive') && yfmJson.archive == true;
        if(archive) {
            console.log("Excluded from panini: ", file.path);
        }
        return archive;
    }))
    .pipe(panini({
      root: 'src/pages',
      layouts: 'src/layouts',
      partials: 'src/partials',
      helpers: 'src/helpers'
    }))
    .pipe(inky())
    .pipe(gulp.dest('dist'));
}
```

edits: variable/keys were wrong

**Comment by ghost** [comment]
@MyKungFuIsGood you'll have to forgive me as I'm a little new to making development updates. 

In the function where the if statement resides, the console.log("Excluded from panini:", file.path); 
do I need to modify any of that (specifically the file.path)? 

As I understand this solution, I wouldn't need to create an archive folder or anything, simply adding the front matter to my files that I want to archive and setting the value to true will keep them from inlining?

**Comment by ben-w-smith** [comment]
@coreyschaaf No worries, the console.log statement does not need to be modified at all. It is there to print out in your terminal as a reminder that some things are being excluded from your directories. You are also correct that you do not need an archive folder or anything else, simply the yaml front matter to your file. 
**_Please note: I only foresaw myself using this on files in pages and don't know what will happen if you tried to use this on partials or layouts**_

--- To better read my code so you know what you are adding:
Gulp uses vinyl which is a virtual file format. Sounds comlicated, gulp uses an object to pass basic information about a file through it's pipes (streams). A vinyl object looks like:

```
var File = require('vinyl');

var coffeeFile = new File({
  cwd: "/",
  base: "/test/",
  path: "/test/file.coffee",
  contents: new Buffer("test = 123")
});
```

So if we were to do console.log(coffeeFile.path) you would get /test/file.coffee

In my additions I'm using $.ignore.exclude to pass the vinyl object to a function that give the object the variable name 'file'. So when you see file.contents or file.path what we're saying is "For this object give me the value for this key."

**Comment by ghost** [comment]
@MyKungFuIsGood sorry it's taken me a bit to get back to you.  Using Front Matter on my file:

```
---
Archive: True
---
```

Does this mean when I use foundation build (or watch?) command that when the inlining process starts the dist folder will no longer be deleted? Or does this mean the dist folder will be deleted, however, it won't inline any files that have the Front Matter 

```
---
Archive: True
---
```

**Comment by ben-w-smith** [comment]
No worries @coreyschaaf. I've only used my implementation with lower case:

```
---
subject: Webinar Template
archive: true
---

<container class="body-content">
```

This will function with all the build commands that already exist. So if you do `$ npm start` or `$ npm run build` these will function as they did before, with the exception that my Webinar Template from pages will not be in the dist folder.

The above code is from `webinar-2016-2-.html` you can see in the image below after running `$ npm run litmus` the dist folder does not have a file for `webinar-2016-6-2.html` in it. This is how it should work for all npm options.

http://i.imgur.com/cPgbAs7.png

**Comment by samumist** [comment]
@MyKungFuIsGood Thanks for the solution. I tried your code but got the following errors when I run `npm start`.

```
[00:47:22] Requiring external module babel-register
[00:47:27] Using gulpfile /Volumes/Data/project/email/elf/gulpfile.babel.js
[00:47:27] Starting 'default'...
[00:47:27] Starting 'build'...
[00:47:27] Starting 'clean'...
[00:47:27] Finished 'clean' after 4.14 ms
[00:47:27] Starting 'pages'...
[00:47:27] 'pages' errored after 11 ms
[00:47:27] TypeError: Cannot read property 'exclude' of undefined
    at pages (gulpfile.babel.js:50:9)
    at bound (domain.js:287:14)
    at runBound (domain.js:300:12)
    at asyncRunner (/Volumes/Data/project/email/elf/node_modules/gulp/node_modules/undertaker/node_modules/bach/node_modules/async-done/index.js:36:18)
    at nextTickCallbackWith0Args (node.js:415:9)
    at process._tickDomainCallback (node.js:385:13)
[00:47:27] 'build' errored after 69 ms
[00:47:27] 'default' errored after 71 ms
```

I just copied and pasted your code into my project's `gulpfile.babel.js` file and `npm installed` the two packages you pointed out in your solution.

I'm using foundation for emails 2.2.1 on Mac OSX 10.11.5 with node 4.3.2 and npm 3.9.5.

Can you please help?

Thanks

**Comment by ghost** [comment]
(Note I'm on a PC) using FFE 2.2.1. 
I just installed this code yesterday and had the same issue: Made the following change from this:

```
  .pipe($.ignore.exclude( function (file) {
```

to this: 

```
   .pipe(ignore.exclude( function (file) {
```

In the same file, navigate to the top of the file. You'll notice some import statements. Add the following to the end of those statements. 

```
import ignore   from 'gulp-ignore';
import yfm      from 'yaml-front-matter';
```

Save the file and then open the command prompt. Navigated to the root of the project folder and ran the following commands. 

```
npm install yaml-front-matter -D
```

(note the -D adds a dependency into your package.json file under dependencies) This way if you ever have to update foundation for emails your package.json file will re-install the dependencies needed for archiving)

After you've installed that, run the following command

```
npm install yaml-front-matter -D
```

Next 
add the front matter to the top of your file you do not want inlined. (I believe case matters so stay lowercase. 

```
archive: true 
```

Next, run the command

```
foundation build
```

The file you added the archive front matter to should not appear is the dist folder. To catch an error you can add the following to the function pages: 

```
console.log(file.path);
```

This will let you know of which file an error is happening on when it's going through the inline process. 

it would look like this for reference: 

```
    .pipe(ignore.exclude( function (file) {
        // parse the file and get a json
        console.log(file.path); 
        var yfmJson = yfm.loadFront(file.contents.toString('ascii'));
        // checks if json has archive key and if its value is true
        var archive = yfmJson.hasOwnProperty('archive') && yfmJson.archive == true;
        if(archive) {
            console.log("Archived: ", file.path);
        }
        return archive;
    }))
```

**Comment by samumist** [comment]
@coreyschaaf 
Thanks very much for your detailed explanation. I followed step by step and it worked as expected for both `npm start` and `npm run build` command. 

As a side note, it reports the following error message the first time I run it.

```
YAMLException: expected valid JSON character
```

that was because I have some Chinese characters in the `front-matter`

```
layout: default
title: 最新讯息
subject: marketing
description: Click to check out the latest news and files for the hottest gigs this summer!
build: production
archive: true
```

I removed the `'ascii'` part in the `yfmJson` declaration and it worked without problem.

Really appreciate your help.
